gcm/sub.go

```diff
@@ -0,0 +1,347 @@
+	"github.com/smancke/guble/store"
+	"math"
+	"strconv"
+	"strings"
+)
+
+const (
+	// default subscription channel buffer size
+	subBufferSize = 50
+)
+
+var (
+	errSubReplaced = errors.New("Subscription replaced")
+)
+
+type jsonError struct {
+	json string
+}
+
+func (e *jsonError) Error() string {
+	return e.json
+}
+
+// Creates a subscription and returns the pointer
+func newSub(gcm *Connector, route *server.Route, lastID uint64) *sub {
+	return &sub{
+		gcm:    gcm,
+		route:  route,
+		lastID: lastID,
+		logger: logger.WithFields(log.Fields{
+			"gcmID":  route.ApplicationID,
+			"userID": route.UserID,
+			"topic":  string(route.Path),
+		}),
+	}
+}
+
+// creates a subscription and adds it in router/kvstore then starts listening for messages
+func initSub(gcm *Connector, topic, userID, gcmID string, lastID uint64) (*sub, error) {
+	route := server.NewRoute(topic, gcmID, userID, subBufferSize)
+	s := newSub(gcm, route, 0)
+	if err := s.store(); err != nil {
+		return nil, err
+	}
+	s.start()
+	return s, nil
+}
+
+// sub represent a GCM subscription
+type sub struct {
+	gcm    *Connector
+	route  *server.Route
+	lastID uint64 // Last sent message id
+
+	logger *log.Entry
+}
+
+func (s *sub) subscribe() error {
+	if _, err := s.gcm.router.Subscribe(s.route); err != nil {
+		s.logger.WithField("error", err).Error("Error subscribing")
+		return err
+	}
+
+	s.logger.Debug("Subscribed")
+	return nil
+}
+
+// unsubscribe from router and remove KVStore
+func (s *sub) remove() *sub {
+	s.gcm.router.Unsubscribe(s.route)
+	s.gcm.kvStore.Delete(schema, s.route.ApplicationID)
+	return s
+}
+
+// start loop to receive messages from route
+func (s *sub) start() error {
+	if err := s.subscribe(); err != nil {
+		return err
+	}
+	go s.subscriptionLoop()
+	return nil
+}
+
+// recreate the route and resubscribe
+func (s *sub) restart() error {
+	s.route = server.NewRoute(string(s.route.Path), s.route.ApplicationID, s.route.UserID, subBufferSize)
+
+	// fetch from last id if applicable
+	if err := s.fetch(); err != nil {
+		return err
+	}
+
+	// subscribe to the router and start
+	return s.start()
+}
+
+// subscriptionLoop that will run in a goroutine and pipe messages from route to gcm
+// Attention: in order for this loop to finish the route channel must be closed
+func (s *sub) subscriptionLoop() {
+	s.logger.Debug("Starting subscription loop")
+
+	s.gcm.wg.Add(1)
+	defer func() {
+		s.logger.Debug("Stopped subscription loop")
+		s.gcm.wg.Done()
+	}()
+
+	// no need to wait for `*gcm.stopC` the channel will be closed by the router anyway
+	for m := range s.route.MessagesChannel() {
+		if err := s.pipe(m); err != nil {
+			// abbandon route if the following 2 errors are met
+			// the subscription has been replaced
+			if err == errSubReplaced {
+				return
+			}
+			// the subscription is not registered with GCM anymore
+			if _, ok := err.(*jsonError); ok {
+				return
+			}
+
+			s.logger.WithField("error", err).Error("Error pipelining message")
```

Comment:
why not use "err" as the message field ? Almost everywhere we used "err" and not error